launcher, converter: Extract Graphics & Video device configuration logic

dasionov opened the pull request:
### What this PR does

The logic for converting a VMI spec to libvirt domain XML is large and complex.

This PR extracts the part responsible for configuring graphics (VNC) and video devices into a dedicated GraphicsDomainConfigurator, improving readability and maintainability.

### References

Partially-Fix: https://github.com/kubevirt/kubevirt/issues/16117

~~Depends-On: #16103~~

### Release note
```release-note
none
```

dasionov commented:
/test pull-kubevirt-check-unassigned-tests

dasionov commented:
[Change](https://github.com/kubevirt/kubevirt/compare/b3d5f6c9c0c80ae95ce262ff69d55e1171838684..2df05996fdfd60d9ef86fa78cd705cb92b608f83):Rebase+Clean

Refactor first commit to use `compute.NewGraphicsDomainConfigurator(architecture, c.BochsForEFIGuests),`

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+func (g GraphicsDomainConfigurator) configureVNCSocket(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
+	domain.Spec.Devices.Graphics = []api.Graphics{
```
Does the order of elements defined in `domain.Spec.Devices.Graphics` change following this commit?

dasionov commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,138 @@
+			vmi.Spec.Domain.Devices.AutoattachGraphicsDevice = autoAttach
+
+			domain := api.Domain{}
+			configurator := compute.NewGraphicsDomainConfigurator("amd64", false)
```
FIXED, Thanks !

dasionov commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -0,0 +1,143 @@
+		}
+		domain.Spec.Devices.Video = []api.Video{video}
+		return nil
+	}
```
fixed

0xFelix commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,138 @@
+			vmi := libvmi.New(libvmi.WithVideo("virtio"))
+			var domain api.Domain
+
+			configurator := compute.NewGraphicsDomainConfigurator(arch, false)
```
Makes sense to test with `bochsForEFIGuests == true` as well? It should not override the user provided value.

dasionov replied:
sure

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -33,5 +35,23 @@ func NewGraphicsDomainConfigurator() GraphicsDomainConfigurator {
 	return nil
 }
+
+func (g GraphicsDomainConfigurator) configureVNCSocket(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
```
This method doesn't seem to access any instance state from `GraphicsDomainConfigurator`. Is there a specific reason to keep it as a method, or could we make a plain function?

dasionov replied:
I understand how you think..  and I agree with you :+1: .
Should I transform it to private helper or keep the few lines inside the Configure func?

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -24,13 +24,26 @@ import (
-type GraphicsDomainConfigurator struct{}
+const (
+	graphicsDeviceDefaultHeads uint = 1
+	graphicsDeviceDefaultVRAM  uint = 16384
```
Are these the defaults for all architectures?

dasionov replied:
currently yes

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -55,3 +69,34 @@ func (g GraphicsDomainConfigurator) configureVNCSocket(vmi *v1.VirtualMachineIns
+
+func (g GraphicsDomainConfigurator) configureAMD64VideoDevice(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
+	// For AMD64 + EFI, use bochs. For BIOS, use VGA
+	if vmi.IsBootloaderEFI() && g.bochsForEFIGuests {
```
The original code evaluated these conditions in reverse order which prevented a redundant function call in case `bochsForEFIGuests` is `false`.

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -71,6 +71,18 @@ func (g GraphicsDomainConfigurator) configureVNCSocket(vmi *v1.VirtualMachineIns
 }
 
 func (g GraphicsDomainConfigurator) configureVideoDevice(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
+	if vmi.Spec.Domain.Devices.Video != nil {
```
In the original code, this logic came after the architecture dependent logic.
Please keep the original order.

dasionov commented:
[Change](https://github.com/kubevirt/kubevirt/compare/1f10e65f6da217e6754bb44aeeec8c6ab5ab2f23..5c37c3a6ff22efd694c6923ac61e6835c9ec4ffd): Squahsed commits 1 and 2 together, fixed commit message signuture, and removed vnc helper func

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,141 @@
+	Context("AutoattachGraphicsDevice", func() {
+
+		DescribeTable("should configure both video and VNC socket based on AutoattachGraphicsDevice", func(arch string, autoAttach *bool) {
+			vmi := libvmi.New(libvmi.WithUID("test-uid"))
```
Is the `WithUID` option required?

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,141 @@
+			Expect(configurator.Configure(vmi, &domain)).To(Succeed())
+
+			if autoAttach == nil || *autoAttach {
+				Expect(domain.Spec.Devices.Video).ToNot(BeEmpty())
```
I recommend asserting the actual domain is equal to an expected domain object.
It makes the behavior more explicit and the test is more sensitive to unexpected changes.

Same bellow.

dasionov replied:
Technically no.
The vnc socket path makes use of the vmi.UID which i find it more readable when configured.

dasionov commented on pkg/virt-launcher/virtwrap/converter/compute/graphics.go:
```diff
@@ -97,3 +99,15 @@ func (g GraphicsDomainConfigurator) configureAMD64VideoDevice(vmi *v1.VirtualMac
 	}
 }
+
+func (g GraphicsDomainConfigurator) configureARM64VideoDevice(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
```
removed

dasionov commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,141 @@
+			} else {
+				Expect(domain.Spec.Devices.Video).To(BeEmpty())
+				Expect(domain.Spec.Devices.Graphics).To(BeEmpty())
+			}
```
done

orelmisan commented on pkg/virt-launcher/virtwrap/converter/arch/BUILD.bazel:
```diff
@@ -11,7 +11,6 @@ go_library(
     importpath = "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/converter/arch",
     visibility = ["//visibility:public"],
     deps = [
-        "//pkg/pointer:go_default_library",
```
nit: Commit message's subject is too long.

dasionov replied:
done

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,381 @@
+			Entry("on s390x", "s390x"),
+		)
+
+		DescribeTable("should configure video and VNC when AutoattachGraphicsDevice is true", func(arch string) {
```
Can this test be united with the `when AutoattachGraphicsDevice is nil`? They should be the same.

dasionov replied:
yes it can, done

dasionov commented on pkg/virt-launcher/virtwrap/converter/arch/arm64.go:
```diff
@@ -21,7 +21,6 @@ import (
 
 	v1 "kubevirt.io/api/core/v1"
 
-	"kubevirt.io/kubevirt/pkg/pointer"
```
done

orelmisan commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,272 @@
+
+var _ = Describe("Graphics Domain Configurator", func() {
+
+	newExpectedAMD64VideoDevice := func() api.Video {
```
Since this is not a closure, I recommend having it a a regular function.
Same bellow.

dasionov commented on pkg/virt-launcher/virtwrap/converter/compute/graphics_test.go:
```diff
@@ -0,0 +1,272 @@
+			Expect(domain).To(Equal(expectedDomain))
+		})
+
+		DescribeTable("should default to virtio", func(arch string, bochsForEFI bool) {
```
yep this is a duplication now, oops.
